### PR TITLE
Adds the Aurora Caelus event, the deep-space equivalent of the northern lights.

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
add: The Nanotrasen Meteorology Division has identified the aurora caelus in your sector. If you are lucky, you may get a chance to witness it with your own eyes.
/:cl:

In response to my steamed hams PR, someone mentioned that it would be interesting if I were to add an aurora borealis... so I did.

On some occasions, a cloud of harmless ions will impact and wrap around the station. What this results in is a rare and breathtaking display of gentle green-and-blue lights around its hull. Nanotrasen's meteorology division approves and encourages staff to take a small break to observe this event, as it may be a once-in-a-lifetime occurrence.

When the aurora ends, starlight will fade back to normal.

This includes music that plays to anyone with midi prefs enabled. The music itself is [Music for Manatees](https://www.youtube.com/watch?v=p-1GtiLcItA) by Kevin Macleod, and is licensed under the CC-BY 3.0 license. It has been shortened to just under two minutes for in-game use. Proper attributions have been included.

*The screenshot does not fully accurately convey its appearance.*

### CODE REVIEW

**Comment by Ordonis** [comment]
Localized ENTIRELY in your station? 

May I see it?

**Comment by Xhuis** [comment]
Yes!

File: code/modules/events/aurora_locus.dm [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/datum/round_event_control/aurora_caelus
+	name = "Aurora Caelus"
+	typepath = /datum/round_event/aurora_caelus
+	max_occurrences = 1
+	weight = 15
+	earliest_start = 5 MINUTES
+
+/datum/round_event_control/aurora_caelus/canSpawnEvent(players, gamemode)
+	if(!CONFIG_GET(flag/starlight))
+		return
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: FALSE

File: code/modules/events/aurora_locus.dm [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	var/aurora_progress = 0 //this cycles from 1 to 7, slowly changing colors from gentle green to gentle blue
+
+/datum/round_event/aurora_caelus/announce()
+	priority_announce("[station_name()]: A harmless cloud of ions is approaching your station, and will exhaust their energy battering the hull. \
+	Nanotrasen has approved a short break for all employees to relax and observe this very rare event. During this time, starlight will be bright \
+	but gentle, shifting between quiet green and blue colors. Any staff who would like to view these lights for themselves may proceed to the area \
+	nearest to them with viewing ports to open space. We hope you enjoy the lights.",
+	sound = 'sound/misc/notice2.ogg',
+	sender_override = "Nanotrasen Meteorology Division")
+	for(var/mob/M in GLOB.player_list)
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: typeless loop

File: code/modules/events/aurora_locus.dm [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+/datum/round_event/aurora_caelus
+	announceWhen = 1
+	startWhen = 10
+	endWhen = 50
+	var/list/aurora_colors = list("#A2FF80", "#A2FF8B", "#A2FF96", "#A2FFA5", "#A2FFB6", "#A2FFC7", "#A2FFDE")
+	var/aurora_progress = 0 //this cycles from 1 to 7, slowly changing colors from gentle green to gentle blue
+
+/datum/round_event/aurora_caelus/announce()
+	priority_announce("[station_name()]: A harmless cloud of ions is approaching your station, and will exhaust their energy battering the hull. \
+	Nanotrasen has approved a short break for all employees to relax and observe this very rare event. During this time, starlight will be bright \
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: No splitting mid string allowed

File: code/modules/events/aurora_locus.dm [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			M.playsound_local(M, 'sound/ambience/aurora_caelus.ogg', 20, FALSE, pressure_affected = FALSE)
+
+/datum/round_event/aurora_caelus/start()
+	for(var/area in GLOB.sortedAreas)
+		var/area/A = area
+		if(initial(A.dynamic_lighting) == DYNAMIC_LIGHTING_IFSTARLIGHT)
+			for(var/turf/open/space/S in A)
+				S.light_range *= 3
+				S.light_power *= 0.5
+				S.update_light()
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: Use set_light

File: code/modules/events/aurora_locus.dm [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+/datum/round_event/aurora_caelus/tick()
+	if(activeFor % 5 == 0)
+		aurora_progress++
+		var/aurora_color = aurora_colors[aurora_progress]
+		for(var/area in GLOB.sortedAreas)
+			var/area/A = area
+			if(initial(A.dynamic_lighting) == DYNAMIC_LIGHTING_IFSTARLIGHT)
+				for(var/turf/open/space/S in A)
+					S.light_color = aurora_color
+					S.update_light()
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: ditto

File: code/modules/events/aurora_locus.dm [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	sender_override = "Nanotrasen Meteorology Division")
+	for(var/mob/M in GLOB.player_list)
+		if((M.client.prefs.toggles & SOUND_MIDI) && is_station_level(M.z))
+			M.playsound_local(M, 'sound/ambience/aurora_caelus.ogg', 20, FALSE, pressure_affected = FALSE)
+
+/datum/round_event/aurora_caelus/start()
+	for(var/area in GLOB.sortedAreas)
+		var/area/A = area
+		if(initial(A.dynamic_lighting) == DYNAMIC_LIGHTING_IFSTARLIGHT)
+			for(var/turf/open/space/S in A)
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: Potentially laggy

File: code/modules/events/aurora_locus.dm [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+				S.update_light()
+
+/datum/round_event/aurora_caelus/tick()
+	if(activeFor % 5 == 0)
+		aurora_progress++
+		var/aurora_color = aurora_colors[aurora_progress]
+		for(var/area in GLOB.sortedAreas)
+			var/area/A = area
+			if(initial(A.dynamic_lighting) == DYNAMIC_LIGHTING_IFSTARLIGHT)
+				for(var/turf/open/space/S in A)
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: Potentially laggy

File: code/modules/events/aurora_locus.dm [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+					S.light_color = aurora_color
+					S.update_light()
+
+/datum/round_event/aurora_caelus/end()
+	for(var/area in GLOB.sortedAreas)
+		var/area/A = area
+		if(initial(A.dynamic_lighting) == DYNAMIC_LIGHTING_IFSTARLIGHT)
+			for(var/turf/open/space/S in A)
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: Potentially laggy

File: code/modules/events/aurora_locus.dm [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+	return to your workplace and continue work as normal. Have a pleasant shift, [station_name()], and thank you for watching with us.",
+	sound = 'sound/misc/notice2.ogg',
+	sender_override = "Nanotrasen Meteorology Division")
+
+/datum/round_event/aurora_caelus/proc/fade_to_black(turf/open/space/S)
+	set waitfor = FALSE
+	var/new_light = initial(S.light_range)
+	while(S.light_range > new_light)
+		S.light_range = max(S.light_range - 0.2, new_light)
+		S.update_light()
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: set_light

File: code/modules/events/aurora_locus.dm [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+		sleep(30)
+	S.light_color = initial(S.light_color)
+	S.light_power = initial(S.light_power)
+	S.update_light()
```
Comment (Cyberboss) [inline code/modules/events/aurora_locus.dm]: ditto

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by Plapatin** [comment]
Well, Xhuis, you are an odd fellow, but I must say, you code a good ham.

File: code/modules/events/aurora_locus.dm (hunk 6 above)
Comment (Xhuis) [reply code/modules/events/aurora_locus.dm]: I tested it locally, and had no issues. What would you recommend instead?

File: code/modules/events/aurora_locus.dm (hunk 3 above)
Comment (Xhuis) [reply code/modules/events/aurora_locus.dm]: Since when? :v

**Comment by AdamElTablawy** [comment]
give it a chance to appear in the kitchen.

**Comment by deathride58** [comment]
@AdamElTablawy that already exists in the form of pyroclastic anomalies

**Comment by PKPenguin321** [comment]
Comfy

**Comment by fludd12** [comment]
Maybe add a sort of gentle weather effect, like how harmless ashfalls on lavaland make a few particles here and there?

**Comment by Anonmare** [comment]
AURORA CAELUS?!
AT THIS TIME OF DAY?!
AT THIS TIME OF YEAR!?
LOCALISED ENTIRELY ON YOUR STATION?!

**Comment by Kromgar** [comment]
It should pulse yellow and purple as well

**Comment by Armhulen** [comment]
this could look way better yeah

**Comment by Xhuis** [comment]
it definitely could, but it's changing a lot of lights at once and I'm trying to minimize the impact it has on lag

**Comment by Cyberboss** [comment]
Up the weight a lot for a test merge

**Comment by Xhuis** [comment]
Try not to let me forget!

**Comment by cebutris** [comment]
Should give pacifism to everyone for the duration of the effect. Nothing like a rare interstellar event to get everyone to set aside their differences and just enjoy life for a little while.

**Comment by Xhuis** [comment]
That seems like it could be something induced by a rare variety that has some sort of anomalous properties. Having it just be "because the lights look rare and pretty" would probably spark some tempers.

**Comment by Cyberboss** [comment]
go ahead and revert it now

**Comment by cebutris** [comment]
The idea came from the music, really. It's really calming and pretty and I don't feel like it would make me want to cave someone's skull in with a fire extinguisher. It's a silly idea anyway and I don't expect it to actually be implemented

**Comment by Xhuis** [comment]
I liked the idea of a version that caused pacifism, so I added that as an admin version and as a 25% chance in general. The pacifism version has a "charged cloud" instead of a harmless one, and will provide a different announce text and give each affected person a message and 5 units of pax.

**Comment by Xhuis** [comment]
Oops. Um...

**Comment by Xhuis** [comment]
Okay, had some git issues but I ironed them out. No pacifism for now!

**Comment by KorPhaeron** [comment]
I have been told this is very laggy

**Comment by Xhuis** [comment]
Local testing hasn't yielded huge amounts of lag. Have there been any testmerges? So far as I'm aware this is done in the smoothest way I can think of, but if it ends up impossible to do smoothly it might be doomed.

**Comment by Nabski89** [comment]
I want to say this was testmerged and it felt like it triggered four rounds in a row.

That might just be because it doesn't have much in the way of requirements for it to trigger?

**Comment by Xhuis** [comment]
Cyberboss asked me to up the weight massively for the testmerge purposes, so I had bumped it up to 100.

**Comment by Dorsisdwarf** [comment]
@Xhuis turn the weighting way down and it might not be rejected for testmerge because "it triggers every single round" which is where this is at currently apparently. the trigger event button isnt just for spamming meteors, cyberboss can use that for testing dont let him bully you

**Comment by Xhuis** [comment]
15 weight is slightly above the average; I set it to 100 earlier but changed it back.

**Comment by ShizCalev** [comment]
@WJohn yay or nay

**Comment by nero1024** [comment]
The .ogg file is a bit rather large, weighting in at 5.4MB. Attached should be a more 'optimized' version with minimal quality loss while at a more reasonable 905KB. Thoughts?
[aurora_caelus_lo.zip](https://github.com/tgstation/tgstation/files/1779204/aurora_caelus_lo.zip)

**Comment by PKPenguin321** [comment]
nero, cant look at that at the moment but i'm assuming you set it to ogg with 0 quality? if you havent already you should also do stereo to mono

**Comment by nero1024** [comment]
It actually was set to 1 quality and resampled to 37.8KHz Stereo. I'm trying to minimize any noticeable quality loss and setting the .ogg to mono will result in too much quality loss, what with this being background music and all.

**Comment by optimumtact** [comment]
why would you set it to mono?

**Comment by nero1024** [comment]
@optimumtact Think of the space savings! However I'm not that crazy to when it comes to chasing for the smallest filesize possible.

**Comment by PKPenguin321** [comment]
mono reduces the filesize by a fuckton and unless youre wearing headphones and have heard the song in stereo before you likely wont notice but i guess it doesnt matter thaaaat much

**Comment by Dorsisdwarf** [comment]
to be fair we have like 5 2Mb songs in the code just for the disco machine dont we?

**Comment by nero1024** [comment]
@Dorsisdwarf Wanna know what's even crazier, boogie2.ogg and superior.ogg are exactly the same, bit for bit. At least these songs have the benefit of being super compressed and stuff.

**Pull request merged**